### PR TITLE
ClusterAPI Provider: Provide fake proivder IDs for failed Machines

### PR DESCRIPTION
**Pull request by JoelSpeed**
This is based on the spot instance handling from the AWS provider (#2235).

Some ClusterAPI implementations will mark a Machine as failed if there is a problem creating the instance on the cloud provider. In this case, these instances weren't being considered as bad and the autoscaler considered them to be `comingUp`. With this patch in, after the `maxNodeProvisionTime` is reached, the autoscaler marks the nodegroup as unhealthy, scales it down (removing the unhealthy machine), and then tries to scale up an alternate group, matching the behaviour seen with the AWS provider once that was patched

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 1]
```diff
@@ -422,6 +443,15 @@ func (c *machineController) machineSetProviderIDs(machineSet *MachineSet) ([]str
 			continue
 		}
 
+		if machine.Status.Phase == "Failed" {
```
Comment (enxebre) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go]: may be compare against ErrorMessage instead?

**Comment by enxebre** [comment]
nit: do you think we can make conflate `Use FailureMessage instead of Phase to determine failed Machines` with the original commit? Tracking that change seems meaningless since it never made it in.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 2]
```diff
@@ -422,6 +443,15 @@ func (c *machineController) machineSetProviderIDs(machineSet *MachineSet) ([]str
 			continue
 		}
 
+		if machine.Status.FailureMessage != nil {
+			klog.V(4).Infof("Status.FailureMessage of machine %q is %q", machine.Name, *machine.Status.FailureMessage)
+			// Provide a fake ID that can be recognised later and converted into a machine key.
```
Comment (enxebre) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:448]: nit: may be add the "why" i.e so the machine is tracked by the nodegroup and the maxNodeProvisionTime can kicks in

**Comment by JoelSpeed** [comment]
@enxebre I addressed your feedback, PTAL

**Pull request merged**